plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -830,6 +840,111 @@ public void rollback()
+
+        Map<ConnectorExpression, HiveApplyProjectionUtil.ProjectedColumnRepresentation> columnProjections = projectedExpressions.stream()
+                .collect(toImmutableMap(Function.identity(), HiveApplyProjectionUtil::createProjectedColumnRepresentation));
+
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) handle;
+
+        // all references are simple variables
+        if (columnProjections.values().stream().allMatch(HiveApplyProjectionUtil.ProjectedColumnRepresentation::isVariable)) {
+            Set<ColumnHandle> projectedColumns = ImmutableSet.copyOf(assignments.values());
+            if (icebergTableHandle.getProjectedColumns().isPresent()
+                    && icebergTableHandle.getProjectedColumns().get().equals(projectedColumns)) {
+                return Optional.empty();
+            }
+            List<Assignment> assignmentsList = assignments.entrySet().stream()
+                    .map(assignment -> new Assignment(
+                            assignment.getKey(),
+                            assignment.getValue(),
+                            ((IcebergColumnHandle) assignment.getValue()).getType()))
+                    .collect(toImmutableList());
+            IcebergTableHandle newTableHandle = new IcebergTableHandle(
+                    icebergTableHandle.getSchemaName(),
+                    icebergTableHandle.getTableName(),
+                    icebergTableHandle.getTableType(),
+                    icebergTableHandle.getSnapshotId(),
+                    icebergTableHandle.getUnenforcedPredicate(),
+                    icebergTableHandle.getEnforcedPredicate(),
+                    Optional.of(projectedColumns));
+
+            return Optional.of(new ProjectionApplicationResult<>(
+                    newTableHandle,
+                    projections,
+                    assignmentsList,
+                    false));
+        }
+
+        Map<String, Assignment> newAssignments = new HashMap<>();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableSet.Builder<ColumnHandle> projectedColumnsBuilder = ImmutableSet.builder();
+
+        for (Map.Entry<ConnectorExpression, HiveApplyProjectionUtil.ProjectedColumnRepresentation> entry : columnProjections.entrySet()) {
+            ConnectorExpression expression = entry.getKey();
+            HiveApplyProjectionUtil.ProjectedColumnRepresentation projectedColumn = entry.getValue();
+
+            IcebergColumnHandle projectedColumnHandle;
+            String projectedColumnName;
+
+            // Create a new column handle
+            IcebergColumnHandle baseColumnHandle = (IcebergColumnHandle) assignments.get(projectedColumn.getVariable().getName());
+            projectedColumnHandle = createProjectedColumnHandle(baseColumnHandle, projectedColumn.getDereferenceIndices(), expression.getType());
+            projectedColumnName = projectedColumnHandle.getQualifiedName();
```

Comment:
`projectedColumnName` will cotain dots. Not sure if that matters. What are we doing for Hive. Is that established pattern?

Reply:
> What are we doing for Hive?

Hive doesn't keep track of the projected columns in the same way. Instead of a path it keeps track of the base column and a list of integer deref indexes to drill down to the desired column. 

> Not sure if that matters

I'm not sure either. For the tests I ran, it didn't, but we could also switch them to a different symbol.

Reply:
No - if nothing breaks I assume this is fine.